contracts/PriceFeed/PriceFeed.sol

```diff
@@ -101,22 +141,54 @@ contract PriceFeed is IPriceFeed, Initializable, ProxyOwned {
         AggregatorV2V3Interface aggregator = aggregators[currencyKey];
-        require(address(aggregator) != address(0), "No aggregator exists for key");
-
-        // this view from the aggregator is the most gas efficient but it can throw when there's no data,
-        // so let's call it low-level to suppress any reverts
-        bytes memory payload = abi.encodeWithSignature("latestRoundData()");
-        // solhint-disable avoid-low-level-calls
-        (bool success, bytes memory returnData) = address(aggregator).staticcall(payload);
-
-        if (success) {
-            (, int256 answer, , uint256 updatedAt, ) = abi.decode(returnData, (uint80, int256, uint256, uint256, uint80));
-            return
-                RateAndUpdatedTime({rate: uint216(_formatAggregatorAnswer(currencyKey, answer)), time: uint40(updatedAt)});
+        IUniswapV3Pool pool = pools[currencyKey];
+        require(address(aggregator) != address(0) || address(pool) != address(0), "No aggregator or pool exists for key");
+
+        if (aggregator != AggregatorV2V3Interface(0)) {
+            // this view from the aggregator is the most gas efficient but it can throw when there's no data,
+            // so let's call it low-level to suppress any reverts
+            bytes memory payload = abi.encodeWithSignature("latestRoundData()");
+            // solhint-disable avoid-low-level-calls
+            (bool success, bytes memory returnData) = address(aggregator).staticcall(payload);
+
+            if (success) {
+                (, int256 answer, , uint256 updatedAt, ) = abi.decode(
+                    returnData,
+                    (uint80, int256, uint256, uint256, uint80)
+                );
+                return RateAndUpdatedTime({rate: uint216(_formatAnswer(currencyKey, answer)), time: uint40(updatedAt)});
+            }
+        } else {
+            uint256 answer =_getPriceFromSqrtPrice(_getTwap(address(pool)));
+            return RateAndUpdatedTime({rate: uint216(_formatAnswer(currencyKey, int256(answer))), time: uint40(block.timestamp)});
         }
     }
 
+    function _getTwap(address pool) public view returns (uint160 sqrtPriceX96) {
+        if (twapInterval == 0) {
+            // return the current price if twapInterval == 0
+            (sqrtPriceX96, , , , , , ) = IUniswapV3Pool(pool).slot0();
+        } else {
+            uint32[] memory secondsAgos = new uint32[](2);
+            secondsAgos[0] = twapInterval; // from (before)
+            secondsAgos[1] = 0; // to (now)
+
+            (int56[] memory tickCumulatives, ) = IUniswapV3Pool(pool).observe(secondsAgos);
+
+            // tick(imprecise as it's an integer) to price
+            sqrtPriceX96 = TickMath.getSqrtRatioAtTick(
+                int24((tickCumulatives[1] - tickCumulatives[0]) / twapInterval)
+            );
+        }
+    }
+
+    function _getPriceFromSqrtPrice(uint160 sqrtPriceX96) public pure returns(uint256 priceX96) {
```

Comment:
does this really need to be public?

Reply:
l've left it public for testing purposes, they should be marked as internal.